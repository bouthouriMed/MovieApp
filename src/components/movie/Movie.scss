@import "../../styles/variables.scss";
@import "../../styles/mixins.scss";

.movie-card {
  position: relative;
  min-width: 370px;
  height: 460px;
  border-radius: 12px;
  overflow: hidden;
  cursor: pointer;
  background: $secondary-color;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: transform 0.3s ease;

  &:hover {
    transform: scale(1.05);
  }

  .movie-link {
    display: block;
    width: 100%;
    height: 100%;

    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
      border-radius: 12px;
    }
  }

  .add-btn {
    position: absolute;
    top: auto;
    right: auto;
    width: 38px;
    height: 38px;
    border-radius: 50%;
    background: #e5091478;
    border: none;
    color: $text-color;
    font-size: 1.2rem;
    font-weight: bold;
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    box-shadow: 0 3px 10px rgba(0, 0, 0, 0.4);
    transition: all 0.25s ease;

    &:hover {
      transform: scale(1.15);
      background: lighten($primary-color, 10%);
    }

    &:active {
      transform: scale(0.95);
      box-shadow: 0 2px 6px rgba(0, 0, 0, 0.3);
    }
  }
}
