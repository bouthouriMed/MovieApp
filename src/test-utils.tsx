import React, { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore, Reducer } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import authReducer, { AuthState } from "./store/authSlice";
import watchListReducer, { WatchListState } from "./store/watchListSlice";
import { moviesApi } from "./store/movieSlice";

// --- Helper for renderHook with optional store ---
import {
  renderHook as rtlRenderHook,
  RenderHookOptions,
} from "@testing-library/react";
import { userApi } from "./store/userSlice";

// Explicit reducers type
export const reducers: {
  auth: Reducer<AuthState>;
  watchList: Reducer<WatchListState>;
  [moviesApi.reducerPath]: typeof moviesApi.reducer;
  [userApi.reducerPath]: typeof userApi.reducer;
} = {
  auth: authReducer,
  watchList: watchListReducer,
  [moviesApi.reducerPath]: moviesApi.reducer,
  [userApi.reducerPath]: userApi.reducer,
};

// Test helper
export const createTestStore = (preloadedState = {}) =>
  configureStore({
    reducer: reducers,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(moviesApi.middleware)
        .concat(userApi.middleware),
    preloadedState,
  });

// --- DOM mocks for scrollable elements ---
export const setupDOMMocks = () => {
  const originalOffsetWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    "offsetWidth"
  );
  const originalScrollWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    "scrollWidth"
  );

  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: 100,
    });
    Object.defineProperty(HTMLElement.prototype, "scrollWidth", {
      configurable: true,
      value: 500,
    });
  });

  afterAll(() => {
    if (originalOffsetWidth)
      Object.defineProperty(
        HTMLElement.prototype,
        "offsetWidth",
        originalOffsetWidth
      );
    if (originalScrollWidth)
      Object.defineProperty(
        HTMLElement.prototype,
        "scrollWidth",
        originalScrollWidth
      );
  });
};

// --- Unified render for components ---
interface RenderOptions {
  store?: ReturnType<typeof createTestStore>;
  wrapper?: React.ComponentType<{ children: ReactNode }>;
}

export const render = (
  ui: React.ReactNode,
  { store, ...renderOptions }: RenderOptions = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (store) {
      return (
        <Provider store={store}>
          <MemoryRouter>{children}</MemoryRouter>
        </Provider>
      );
    }
    return <MemoryRouter>{children}</MemoryRouter>;
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

export const renderHookWithStore = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  {
    store,
    wrapper,
    ...options
  }: RenderHookOptions<TProps> & {
    store?: ReturnType<typeof createTestStore>;
  } = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (wrapper) return React.createElement(wrapper, null, children);
    return store ? (
      <Provider store={store}>
        <MemoryRouter>{children}</MemoryRouter>
      </Provider>
    ) : (
      <MemoryRouter>{children}</MemoryRouter> // always wrap in Router
    );
  };
  return rtlRenderHook(hook, { wrapper: Wrapper, ...options });
};

export const renderWithoutRouter = (
  ui: React.ReactNode,
  { store, ...renderOptions }: RenderOptions = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return store ? (
      <Provider store={store}>{children}</Provider>
    ) : (
      <>{children}</>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};
