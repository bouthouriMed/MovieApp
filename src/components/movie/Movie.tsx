import { memo } from "react";
import { Link } from "react-router-dom";
import "./Movie.scss";

interface MovieProps {
  movie: any;
  onAdd?: (movie: any) => void;
  category?: string;
}

function Movie({ movie, category }: MovieProps) {
  return (
    <div className="movie-card">
      <Link
        to={`/movie/${movie.id}`}
        key={movie.id}
        state={{ category }}
        className="movie-link"
      >
        <img
          src={`https://image.tmdb.org/t/p/w300${movie.poster_path}`}
          alt={movie.title}
        />
      </Link>

      {/* {onAdd && (
        <button
          className="add-btn"
          onClick={() => onAdd(movie)}
          title="Add to Watchlist"
        >
          âž•
        </button>
      )} */}
    </div>
  );
}

export default memo(Movie);
