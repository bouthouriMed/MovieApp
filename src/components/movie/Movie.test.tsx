import { render } from "@/test-utils";
import Movie from "./Movie";

const movie = { id: 1, title: "Test Movie", poster_path: "/poster.jpg" };

describe("Movie", () => {
  it("matches snapshot", () => {
    const { container } = render(<Movie movie={movie} category="Action" />);
    expect(container).toMatchSnapshot();
  });
});
