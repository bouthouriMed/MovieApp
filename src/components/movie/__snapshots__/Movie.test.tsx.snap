// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Movie > matches snapshot 1`] = `
<div>
  <div
    class="movie-card"
  >
    <a
      class="movie-link"
      data-discover="true"
      href="/movie/1"
    >
      <img
        alt="Test Movie"
        src="https://image.tmdb.org/t/p/w300/poster.jpg"
      />
    </a>
  </div>
</div>
`;
